### ISSUE TITLE
Cannot restore unix socket of X11: No such file or directory

### ISSUE THREAD

**Issue by muvaf** [checkpoint-restore/criu#2359]
**Description**

I have a VNC server and a browser running inside a container. Dumping works fine, but when I restore, it fails saying that it cannot find `/tmp/.X11-unix/X0`, however, `Xtigervnc` uses that file as its display socket and I believe the browser communicates through that and all processes are included in the process tree. So, it should be able to restore as both ends of the socket are included.

**Describe the results you received:**

```
Warn  (criu/kerndat.c:1593): CRIU was built without libnftables support
Warn  (criu/kerndat.c:1243): Can't keep kdat cache on non-tempfs
  9296: Error (criu/sk-unix.c:1651): unix: Can't bind id 0x81 ino 75376689 addr /tmp/.X11-unix/X0: No such file or directory
  9296: Error (criu/files.c:1213): Unable to open fd=5 id=0x81
  9289: Error (criu/cr-restore.c:1513): 9296 exited, status=1
Error (criu/cr-restore.c:2557): Restoring FAILED.
```

**Describe the results you expected:**

Restored and continuing to work. I don't expect the TCP connections to continue from where they were since IP of container does change, hence `--tcp-close`

**Additional information you deem important (e.g. issue happens only occasionally):**

I'm using swrk mode with a wrapper to take the checkpoint but CLI to restore. The options boil down to the following:

<details><summary>Checkpoint options:</summary>
<p>

```
		TcpEstablished:  proto.Bool(true),
		ShellJob:        proto.Bool(false),
		FileLocks:       proto.Bool(false),
		LogFile:         proto.String("dump.log"),
		AutoDedup:       proto.Bool(false),
		Pid:             proto.Int32(int32(pid)),
		ImagesDirFd:     proto.Int32(int32(fd)), // To make it use ImagesDir.
		OrphanPtsMaster: proto.Bool(true),
		NotifyScripts:   proto.Bool(true),
		LeaveRunning:    proto.Bool(false),
		LogLevel:        proto.Int32(4),
		LazyPages:       proto.Bool(false),
		GhostLimit:      proto.Uint32(500 * 1048576), // 500MB
		Root:            proto.String("/"),
		TcpClose:        proto.Bool(true),
		ExtMnt: []*rpc.ExtMountMap{
		{
			Key: proto.String("/proc/kcore"),
			Val: proto.String("/dev/null"),
		},
		{
			Key: proto.String("/proc/keys"),
			Val: proto.String("/dev/null"),
		},
		{
			Key: proto.String("/proc/timer_list"),
			Val: proto.String("/dev/null"),
		},
		{
			Key: proto.String("/dev/null"),
			Val: proto.String("/dev/null"),
		},
		{
			Key: proto.String("/dev/random"),
			Val: proto.String("/dev/random"),
		},
		{
			Key: proto.String("/dev/full"),
			Val: proto.String("/dev/full"),
		},
		{
			Key: proto.String("/dev/zero"),
			Val: proto.String("/dev/zero"),
		},
		{
			Key: proto.String("/dev/urandom"),
			Val: proto.String("/dev/urandom"),
		},
		{
			Key: proto.String("/dev/tty"),
			Val: proto.String("/dev/tty"),
		},
		{
			Key: proto.String("/usr/share/zoneinfo"),
			Val: proto.String("/usr/share/zoneinfo"),
		},
		{
			Key: proto.String("/sys/fs/cgroup/kubepods.slice/kubepods-burstable.slice"),
			Val: proto.String("/sys/fs/cgroup/kubepods.slice/kubepods-burstable.slice"),
		},
	},
```

</p>
</details>

Restore:
```
criu restore --images-dir /checkpoint --tcp-established --file-locks --evasive-devices --tcp-close -v4 --log-file restore.log --inherit-fd fd[1]:pipe:[75376685] --inherit-fd fd[2]:pipe:[75376686]
```

**CRIU logs and information:**

CRIU full dump/restore logs:
[dump.log](https://github.com/checkpoint-restore/criu/files/14479377/dump.log)
[restore.log](https://github.com/checkpoint-restore/criu/files/14479685/restore.log)

<details><summary>Output of `criu --version`:</summary>
<p>

```
Version: 3.19
```

</p>
</details>

<details><summary>Output of `criu check --all`:</summary>
<p>

```
Warn  (criu/kerndat.c:1593): CRIU was built without libnftables support
Warn  (criu/kerndat.c:1243): Can't keep kdat cache on non-tempfs
Warn  (criu/cr-check.c:822): Dirty tracking is OFF. Memory snapshot will not work.
Warn  (criu/cr-check.c:1257): Do not have API to map vDSO - will use mremap() to restore vDSO
Error (criu/cr-check.c:1223): UFFD is not supported
Error (criu/cr-check.c:1223): UFFD is not supported
Warn  (criu/cr-check.c:1346): Nftables based locking requires libnftables and set concatenations support
Error (criu/cr-check.c:1005): failed to mount autofs: No such device
Warn  (criu/cr-check.c:1177): CRIU built without CONFIG_COMPAT - can't C/R compatible tasks
Looks good but some kernel features are missing
which, depending on your process tree, may cause
dump or restore failure.
```

</p>
</details>

**Additional environment details:**

The processes are started in a `entrypoint.sh` that's wrapped by `procman` process that I wrote that makes criu calls. The reason for high PID numbers is that it bumps the last pid first before starting `entrypoint.sh`. It uses `setsid` and `unshare -i` equivalent of Go code to start it.

**Comment by adrianreber** [comment]
Does the directory `/tmp/.X11-unix` exist before doing the restore?

**Comment by adrianreber** [comment]
Independent of the error, looking at your options, you could probably use cgroupmode ignore.

**Comment by muvaf** [comment]
> Does the directory /tmp/.X11-unix exist before doing the restore?

It didn't and now I added, and it seems to pass. Thank you! Would it make sense to make criu run `mkdir -p` for socket files that may be in subdirectories like in this case? I can create a patch if you can point me to where that happens.

> you could probably use cgroupmode ignore.

What does this do exactly? After getting through the error above, I'm now seeing `10705: Error (criu/cr-restore.c:3071): Inconsistent sched params received (1073741824.0.0)`. Could it be related to that option?

**Comment by adrianreber** [comment]
> > Does the directory /tmp/.X11-unix exist before doing the restore?
> 
> It didn't and now I added, and it seems to pass. Thank you! Would it make sense to make criu run `mkdir -p` for socket files that may be in subdirectories like in this case? I can create a patch if you can point me to where that happens.

Not sure. Usually CRIU avoids file system modifications. In this case I could understand the need for a `mkdir`.

> > you could probably use cgroupmode ignore.
> 
> What does this do exactly?

I just saw that you are mapping cgroup settings with the help of `--external`. If you are running in a container and do not wish to change the cgroup settings you can just tell criu to ignore cgroup and just restore the process without adapting the cgroup settings. As you are running in Kubernetes I would assume you do not want to change cgroup settings in your container but just use what Kubernetes gives you. In that case ignore makes sense.

> After getting through the error above, I'm now seeing `10705: Error (criu/cr-restore.c:3071): Inconsistent sched params received (1073741824.0.0)`. Could it be related to that option?

No. Looking at the checkpoint I see:
```
(00.025140) 9289 has 0 sched policy
(00.025142) 	dumping 0 nice for 9289
```

CRIU verifies that the restored process has the same. That seems to fail.

Although, this looks strange: ` Inconsistent sched params received (1073741824.0.0)` this looks like 2^30 which is a strange policy. So not sure what is going on.

**Comment by muvaf** [comment]
Here is updated logs from the run I'm getting this error.
[dump.log](https://github.com/checkpoint-restore/criu/files/14481122/dump.log)
[restore.log](https://github.com/checkpoint-restore/criu/files/14481124/restore.log)

In dump logs, there are tasks with `1073741824` sched policy but I do see `10704 has 0 sched policy`. In restore, the same task fails with `10704: Error (criu/cr-restore.c:3071): Inconsistent sched params received (1073741824.0.0)`. Unlikely, but wonder if there is a bug somewhere that mixes them up somehow.

By the way, I can close this one and open a new issue for this error if you'd like to since the original issue is resolved.

**Comment by rst0git** [comment]
> In dump logs, there are tasks with 1073741824 sched policy but I do see 10704 has 0 sched policy. In restore, the same task fails with 10704: Error (criu/cr-restore.c:3071): Inconsistent sched params received (1073741824.0.0). Unlikely, but wonder if there is a bug somewhere that mixes them up somehow.

@muvaf Would you be able to create a simple example to reproduce this error?

**Comment by adrianreber** [comment]
@muvaf You can inspect the scheduling policy settings using crit: `crit show core-xxx.img` with the correct process ID:
```
            "thread_core": {
                "futex_rla": 0,
                "futex_rla_len": 24,
                "sched_nice": 0,
                "sched_policy": 0,
```

**Comment by adrianreber** [comment]
Ah, the value is set to `# define SCHED_RESET_ON_FORK    0x40000000` which should be 2^30

**Comment by muvaf** [comment]
@rst0git It's a bit complex but I'll try. It's effectively Xtigervnc, websockify and node process that spawns webkit browser. I'll try to narrow it down for easier reproducibility.

@adrianreber Did another run, there is not even mention of `1073741824` in the output of `crit show core-10696.img` which is the PID show in the error log. 
[core-10696.txt](https://github.com/checkpoint-restore/criu/files/14481735/core-10696.txt)
[dump.log](https://github.com/checkpoint-restore/criu/files/14481730/dump.log)
[restore.log](https://github.com/checkpoint-restore/criu/files/14481731/restore.log)

Another process `10720` does have that policy in the dump logs and I can see it in its `crit show core-10720.img` output as well.
[core-10720.txt](https://github.com/checkpoint-restore/criu/files/14481764/core-10720.txt)

**Comment by adrianreber** [comment]
> > Ah, the value is set to # define SCHED_RESET_ON_FORK    0x40000000 which should be 2^30
> 
> Looks like this is something set in the kernel, not criu? FWIW, my kernel is `6.4.16-linuxkit` with `aarch64`.

Yes, that is from the kernel. It was mainly good to find. So the value is not some overflow or a bug, but a real value the kernel can set.

**Comment by adrianreber** [comment]
The problem probably is that at https://github.com/checkpoint-restore/criu/blob/criu-dev/criu/cr-restore.c#L3060 SCHED_RESET_ON_FORK is not part of the switch statement. But it seems to be a valid value. So that could be a CRIU bug. Not sure why it is missing.

**Comment by adrianreber** [comment]
@avagin do you know anything about SCHED_RESET_ON_FORK and why CRIU does not handle it?

**Comment by rst0git** [comment]
> In dump logs, there are tasks with 1073741824 sched policy but I do see 10704 has 0 sched policy. In restore, the same task fails with 10704: Error (criu/cr-restore.c:3071): Inconsistent sched params received (1073741824.0.0). Unlikely, but wonder if there is a bug somewhere that mixes them up somehow.

@muvaf I was able to replicate this error locally with the following example and opened a pull request with a fix (https://github.com/checkpoint-restore/criu/pull/2360):
```C
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <sched.h>
#include <unistd.h>
#include <errno.h>

int main()
{
	struct sched_param param;
	param.sched_priority = 99;

	if (sched_setscheduler(getpid(), SCHED_FIFO | SCHED_RESET_ON_FORK, &param) == -1) {
		perror("Error setting scheduler");
		exit(EXIT_FAILURE);
	}

	int policy = sched_getscheduler(getpid());
	if (policy == -1) {
		perror("Error getting scheduler");
		exit(EXIT_FAILURE);
	}

	if (policy != (SCHED_FIFO | SCHED_RESET_ON_FORK)) {
		printf("Failed to set scheduling policy: %d != %d\n",
		       policy, SCHED_FIFO | SCHED_RESET_ON_FORK);
		exit(EXIT_FAILURE);
	}

	while (1)
		sleep(1);

	return 0;
}
```

**Comment by muvaf** [comment]
@rst0git I built your PR and tested it in my setup. I don't see that sched error anymore and finally I've seen some processes getting started, thank you!

I got another error similar to the one here https://github.com/checkpoint-restore/criu/issues/2072 ([restore.log](https://github.com/checkpoint-restore/criu/files/14494076/restore.log)) though with CLI call of `criu restore --images-dir /checkpoint --tcp-established --file-locks --evasive-devices --tcp-close --manage-cgroups=ignore -v4 --log-file restore.log --inherit-fd fd[1]:pipe:[82471854] --inherit-fd fd[2]:pipe:[82471855]`. I'm guessing it's likely because of my local build; I built for aarch64 using `make -C scripts/build/ aarch64-stable-cross` and then copied only the binary in `/criu/criu/criu` to override the `/usr/sbin/criu` binary on my Ubuntu 22.04 image that had 3.19 installed through apt.
